### PR TITLE
Fix milestone events for new execution phases

### PR DESCRIPTION
**Pull request by n3wscott**

### CODE REVIEW

File: pkg/environment/magic.go [hunk 1]
```diff
@@ -177,6 +177,11 @@ func (mr *MagicEnvironment) Prerequisite(ctx context.Context, t *testing.T, f *f
 func (mr *MagicEnvironment) Test(ctx context.Context, originalT *testing.T, f *feature.Feature) {
 	originalT.Helper() // Helper marks the calling function as a test helper function.
 
+	mr.milestones.TestStarted(f.Name, originalT.Name())
+	originalT.Cleanup(func() {
+		mr.milestones.TestFinished(f.Name, originalT.Name(), originalT.Skipped(), originalT.Failed())
```
Comment (vaikas) [inline pkg/environment/magic.go:182]: Seeing this called, why not change the signature to:
(name string, originalTest *testing.T)

That way if we ever need anything else from the original T we don't have to change the signature and callers?

File: pkg/environment/magic.go [hunk 2]
```diff
@@ -194,12 +199,17 @@ func (mr *MagicEnvironment) Test(ctx context.Context, originalT *testing.T, f *f
...
-		var internalT *testing.T
+		var internalT feature.T
 		originalT.Run(s.T.String()+"/"+s.TestName(), func(st *testing.T) {
 			internalT = st
 			st.Helper()
 			st.Cleanup(wg.Done) // Make sure wg.Done() is always invoked, no matter what
 
+			mr.milestones.StepStarted(f.Name, s.Name, s.T.String(), s.L.String(), st.Name())
+			originalT.Cleanup(func() {
+				mr.milestones.StepFinished(f.Name, s.Name, s.T.String(), s.L.String(), st.Name(), internalT.Skipped(), internalT.Failed())
```
Comment (vaikas) [inline pkg/environment/magic.go:210]: And the same here:
StepFinished(featureName string, step *Step, origTest *testing.T)

**Review by vaikas — COMMENTED** [review]
Just a thought about the signatures, just curious if they are going to get unwieldy.

File: pkg/environment/magic.go (hunk 2 above)
Comment (n3wscott) [reply pkg/environment/magic.go]: sounds like a good idea!

File: pkg/environment/magic.go (hunk 1 above)
Comment (n3wscott) [reply pkg/environment/magic.go]: good idea!

File: pkg/milestone/emitter.go [hunk 3]
```diff
@@ -45,10 +47,12 @@ type Emitter interface {
 	Environment(env map[string]string)
 	NamespaceCreated(namespace string)
 	NamespaceDeleted(namespace string)
-	TestStarted(feature, stepName, testName string)
-	TestFinished(feature, stepName, testName string, skipped, failed bool)
-	TestSetStarted(featureSet, testName string)
-	TestSetFinished(featureSet, testName string, skipped, failed bool)
+	TestStarted(feature string, t feature.T)
+	TestFinished(feature string, t feature.T)
+	StepStarted(feature string, step *feature.Step, t feature.T)
```
Comment (vaikas) [inline pkg/milestone/emitter.go:52]: zomg one more suggest now that I look at this, sorry :)

feature string, t feature.T, step *feature.Step

Just for consistency of the API? So, featureName first, then Feature, then Step.

Comment (vaikas) [reply pkg/milestone/emitter.go:52]: Ok, I guess feature.Step is more higher level, I had it backwards. LGTM.

**Review by vaikas — APPROVED** [review]

File: test/e2e/environment_fails_test.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	"knative.dev/reconciler-test/pkg/feature"
+)
+
+// These are example tests that fail intentionally to demonstrate feature timings working.
+
+func TestRequirementSkip(t *testing.T) {
```
Comment (slinkydeveloper) [inline test/e2e/environment_fails_test.go:35]: This test should not fail, it should just be skipped

**Pull request merged**